rust/connlib/tunnel/src/io.rs

```diff
@@ -126,15 +126,12 @@ impl Io {
         self.upstream_dns_servers = create_resolvers(dns_servers);
     }
 
-    pub fn perform_dns_query(&mut self, query: DnsQuery<'static>) {
+    pub fn perform_dns_query(&mut self, query: DnsQuery<'static>) -> Result<(), DnsQueryError> {
         let upstream = query.query.destination();
         let Some(resolver) = self.upstream_dns_servers.get(&upstream).cloned() else {
-            tracing::warn!(%upstream, "Dropping DNS query because of unknown upstream DNS server");
-            return;
+            return Err(DnsQueryError::UnknownServer);
```

Comment:
```suggestion
            // This can only happen in a rare race condition where the packet was constructed just before the DNS servers have changed
            return Err(DnsQueryError::UnknownServer);
```

Reply:
I actually think this can never happen: If we get to this point, it means that the DNS query was successfully routed to our TUN interface. At that point, it MUST have been sent to a sentinel DNS, right?

If we get to the point where we are deferring the query to an upstream server (i.e. here), it means we have already mangled the DNS query so the mapping at least exists in `ClientState`. If it doesn't exist here, it means we have inconsistent state.